src/dune_pkg/rev_store.ml

```diff
@@ -2,43 +2,97 @@ open Stdune
 open Dune_vcs
 module Process = Dune_engine.Process
 module Display = Dune_engine.Display
+module Scheduler = Dune_engine.Scheduler
 module Re = Dune_re
+module Flock = Dune_util.Flock
 open Fiber.O
 
-type t = { dir : Path.t }
+type t =
+  { dir : Path.t
+  ; lock : Path.t
+  }
+
 type rev = Rev of string
 
-let equal { dir } t = Path.equal dir t.dir
+(* Async-inspired variation of [Fiber.repeat_until] *)
+let rec repeat_until_finished state f =
+  let* computation = f state in
+  match computation with
+  | `Repeat state -> repeat_until_finished state f
+  | `Finished result -> Fiber.return result
+;;
+
+let lock_limited flock lock timeout =
+  let timeout = max timeout 0. in
+  let sleep_duration = 0.1 in
+  repeat_until_finished timeout (fun timeout ->
+    if timeout > 0.
+    then (
+      match Flock.lock_non_block flock lock with
+      | Ok `Success as ok -> Fiber.return @@ `Finished ok
+      | Ok `Failure ->
+        let+ () = Scheduler.sleep sleep_duration in
+        `Repeat (timeout -. sleep_duration)
+      | err -> Fiber.return @@ `Finished err)
+    else Fiber.return @@ `Finished (Ok `Failure))
+;;
+
+let with_flock lock_path ~f =
+  let open Fiber.O in
+  let parent = Path.parent_exn lock_path in
+  Path.mkdir_p parent;
+  let fd = Unix.openfile (Path.to_string lock_path) [ Unix.O_CREAT; O_RDONLY ] 0o644 in
+  let flock = Flock.create fd in
+  let max_duration = 5. in
+  Fiber.finalize
+    ~finally:(fun () ->
+      (* closing the fd releases the flock automatically *)
+      Unix.close fd;
```

Comment:
What happens if `close` fails? (`close_noerr` is probably good enough?)

Reply:
If `close` fails, it's very likely that it's `EBADFD` and that means we have a bug somewhere. We shouldn't ignore it.

Reply:
I've wrapped it with some error handling code. Unfortunately I don't think I can use the same `strace` trick since injecting into `close` is would require knowing the FD otherwise everything breaks.